libs/pubsub/query/query.go

```diff
@@ -223,8 +223,15 @@ func compileCondition(cond syntax.Condition) (condition, error) {
 // tests for, but we should probably get rid of that.
 var extractNum = regexp.MustCompile(`^\d+(\.\d+)?`)
```

Comment:
If you just type in 12.434CXXD it will fail because there is nothing to parse it. If you type in "12.34CXXD" it will parse it as a string but then fail when comparing it to a number because it will try to parse it as a float.

Reply:
Not sure I get why "12.434CXXD" and "12.34CXXD" will be treated differently.

Anyway, Mm main concern here is that the regex does not require the string to _end_ with digits... i.e., there is no `$` at the end of the regex

Reply:
I had a typo in my previous comment but I did a bit more digging and found a discrepancy between query parsing in the kvindexer and the pubsub query parser. First of all, "12.434CXXD" and "12.34CXXD"  are exactly the same, I mistakenly did not write an extra 4. In pubsub it will convert this to 12.434.In the kvindexer it will fail on conversion to Float/Int. Same for 12.43CXXD. If you quote either one of those, it will parse them as a string and then your comparison is only on equality. 
I wanted to rewrite it initially as you suggested but, I noticed that the test 
`amount > 7` where `amount=8stake`  fails when we add the `$` but passes without it. I know some of our users indeed input amounts like that (`8stake, 8atom` etc). However, the same query on the kvindexer will fail. The reason is that the pubsub basically [parses out the leftmost ](https://github.com/cometbft/cometbft/blob/2e0c72b94f3d817569289d6d6baa284cbbb3619a/libs/pubsub/query/query.go#L227) match on the string. So  out of `12.434CXXD` convert `12.434` to float. I think we should introduce this behaviour in our kvindexer too, but I would to that in a separate PR because it is another module, and  not related to the change here.

Reply:
100% agree. The "8atom" think makes sense and we shouldn't break them just because we don't agree with that. So I agree with making the behavior consistent in the kvindexer and, maybe, replacing the current TODO comment above the regex by a couple of sentences explaining the "8stake, 6atom" case.

Reply:
I just noticed you're off this week. I'll try to merge this today on your behalf